### PR TITLE
Extended URP code improvements.

### PR DESCRIPTION
**Pull request by iccub**

### CODE REVIEW

File: brave/src/analytics/user-referral-program/UserReferralProgram.swift [hunk 1]
```diff
@@ -178,24 +180,21 @@ class UserReferralProgram {
...
-    class func addCustomHeaders(to request: URLRequest) -> URLRequest {
-
+    /// Checks if a custom header should be added to the request and returns its value and field.
+    class func shouldAddCustomHeader(for request: URLRequest) -> (value: String, field: String)? {
         guard let prefs = BraveApp.getPrefs(), let customHeadersAsData: Data = prefs.objectForKey(CustomHeaderData.prefsKey) as Any? as? Data,
-            let customHeaders = NSKeyedUnarchiver.unarchiveObject(with: customHeadersAsData) as? [CustomHeaderData],
-            let hostUrl = request.url?.host else { return request }
-
-        var newRequest = request
+            let customHeaders = NSKeyedUnarchiver.unarchiveObject(with: customHeadersAsData) as? [CustomHeaderData], let hostUrl = request.url?.host else { return nil }
```
Comment (jhreis) [inline brave/src/analytics/user-referral-program/UserReferralProgram.swift]: This on two lines?

File: brave/src/analytics/user-referral-program/UserReferralProgram.swift [hunk 2]
```diff
@@ -178,24 +180,21 @@ class UserReferralProgram {
...
 
         for customHeader in customHeaders {
-            innerLoop: for domain in customHeader.domainList {
+            for domain in customHeader.domainList {
                 if hostUrl.contains(domain) {
-                    UrpLog.log("Adding custom header: [\(customHeader.headerField): \(customHeader.headerValue)] for domain: \(domain)")
-                    newRequest.addValue(customHeader.headerValue, forHTTPHeaderField: customHeader.headerField)
-                    break innerLoop
+                    if let allFields = request.allHTTPHeaderFields, !allFields.keys.contains(customHeader.headerField) {
+                        return (customHeader.headerValue, customHeader.headerField)
```
Comment (jhreis) [inline brave/src/analytics/user-referral-program/UserReferralProgram.swift]: One thing to note, this only will send a single header, old implementation looped through all headers and attached could attach a domain to all of them. Was this intentional?

File: brave/src/webview/BraveWebView.swift [hunk 3]
```diff
@@ -374,16 +374,14 @@ class BraveWebView: UIWebView {
...
-
         guard let internalWebView = value(forKeyPath: "documentView.webView") else { return }
         NotificationCenter.default.removeObserver(self, name: NSNotification.Name(rawValue: internalProgressChangedNotification), object: internalWebView)
         NotificationCenter.default.addObserver(self, selector: #selector(BraveWebView.internalProgressNotification(_:)), name: NSNotification.Name(rawValue: internalProgressChangedNotification), object: internalWebView)
 
         if let url = request.url, let host = url.normalizedHost {
             internalSetBraveShieldStateForDomain(host)
         }
-        super.loadRequest(mutatedRequest)
+        super.loadRequest(request)
```
Comment (jhreis) [inline brave/src/webview/BraveWebView.swift]: This back to default behavior?

Comment (iccub) [reply brave/src/webview/BraveWebView.swift]: We set headers in a different place now, `loadRequest` was in the first implementation.

Comment (jhreis) [reply brave/src/webview/BraveWebView.swift]: Yeah, just confirming 👍

File: brave/src/analytics/user-referral-program/UserReferralProgram.swift (hunk 2 above)
Comment (jhreis) [reply brave/src/analytics/user-referral-program/UserReferralProgram.swift]: Per added comments, 👍

**Pull request merged**